### PR TITLE
Adds method to read the pooling types from model's files

### LINKED ISSUE #9388: [Feature]: Support sentence-transformers configuration files
### 🚀 The feature, motivation and pitch

Currently support for sentence transformer models being added in PRs such as https://github.com/vllm-project/vllm/pull/9056. However, these models come with extra configuration files for settings such as the pooling method and whether normalization of embeddings has to be done.

There is a `modules.json` file as such as [this one](https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2/blob/main/modules.json_ containing configuration of layers:
```
[
  {
    "idx": 0,
    "name": "0",
    "path": "",
    "type": "sentence_transformers.models.Transformer"
  },
  {
    "idx": 1,
    "name": "1",
    "path": "1_Pooling",
    "type": "sentence_transformers.models.Pooling"
  },
  {
    "idx": 2,
    "name": "2",
    "path": "2_Normalize",
    "type": "sentence_transformers.models.Normalize"
  }
]
```

The path refers to a directory that can be empty or non-existent in the case of default parameters. For example, in the case of the model above, `1_Pooling` contains a `config.json` file with the following:
```
{
  "word_embedding_dimension": 384,
  "pooling_mode_cls_token": false,
  "pooling_mode_mean_tokens": true,
  "pooling_mode_max_tokens": false,
  "pooling_mode_mean_sqrt_len_tokens": false
}
```

Currently the implementation of BERT models and Roberta Models hard codes the most common option of these parameters, but to really support sentence transformer models we need to load the settings from these files.

cc: @robertgshaw2-neuralmagic @flaviabeo 

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by flaviabeo**
This adds a method to load the pooling config file from sentence transformer models like [sentence-transformers/all-MiniLM-L12-v2](https://huggingface.co/sentence-transformers/all-MiniLM-L12-v2/blob/main/1_Pooling/config.json).

The pooling types added can be found at the [sentence-transformers Pooling](https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/models/Pooling.py#L19)

FIX #9388 (*link existing issues this PR will resolve*)

cc: @maxdebayser 

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Adding or changing kernels</h3>
<p>Each custom kernel needs a schema and one or more implementations to be registered with PyTorch.</p>
<ul>
    <li>Make sure custom ops are registered following PyTorch guidelines: <a href="https://pytorch.org/tutorials/advanced/cpp_custom_ops.html#cpp-custom-ops-tutorial">Custom C++ and CUDA Operators</a> and <a href="https://docs.google.com/document/d/1_W62p8WJOQQUzPsJYa7s701JXt0qf2OfLub2sbkHOaU">The Custom Operators Manual</a></li>
    <li>Custom operations that return <code>Tensors</code> require meta-functions. Meta-functions should be implemented and registered in python so that dynamic dims can be handled automatically. See above documents for a description of meta-functions.</li>
    <li>Use <a href="https://pytorch.org/docs/stable/library.html#torch.library.opcheck"><code>torch.libary.opcheck()</code></a> to test the function registration and meta-function for any registered ops.  See <code>tests/kernels</code> for examples.</li>
    <li>When changing the C++ signature of an existing op, the schema must be updated to reflect the changes.</li>
    <li>If a new custom type is needed, see the following document: <a href="https://docs.google.com/document/d/18fBMPuOJ0fY5ZQ6YyrHUppw9FA332CpNtgB6SOIgyuA">Custom Class Support in PT2</a>.
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by robertgshaw2-neuralmagic** [comment]
Thanks! This is great! Can you make sure to add some integration tests?

File: vllm/model_executor/models/bert.py [hunk 1]
```diff
@@ -382,15 +382,17 @@ class BertEmbeddingModel(nn.Module):
...
+    def __init__(self,
+                 config: BertConfig,
+                 cache_config: Optional[CacheConfig] = None,
+                 quant_config: Optional[QuantizationConfig] = None,
+                 pooling_config: Optional[PoolingConfig] = None) -> None:
         super().__init__()
         self.model = BertModel(config, cache_config, quant_config)
-        self._pooler = Pooler(pooling_type=PoolingType.CLS, normalize=True)
+        print(pooling_config.pooling_type)
+        print(pooling_config.normalize)
```
Comment (DarkLight1337) [inline vllm/model_executor/models/bert.py:393]: You forgot to delete these from the PR.
```suggestion
```

File: vllm/model_executor/layers/pooler.py [hunk 2]
```diff
@@ -13,6 +13,30 @@ class PoolingType(IntEnum):
     LAST = 0
     ALL = 1
     CLS = 2
+    MEAN = 3
+
+
+class PoolingConfig():
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:19]: Maybe make this a dataclass?

File: vllm/transformers_utils/config.py [hunk 3]
```diff
@@ -232,20 +232,76 @@ def get_config(
...
+    with open(file_path, "r") as file:
+        config_dict = json.load(file)
+
+    return config_dict
+
+
+def get_pooling_config(model, revision='main'):
+    """
+    This function gets the pooling and normalize 
+    config from the model.
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py:263]: Can we rename this so it's more obvious that this only applies to sentence-transformers models? We currently already have embedding models that are not from sentence-transformers, so it's rather confusing why only BERT has `PoolingConfig`.

Comment (flaviabeo) [reply vllm/transformers_utils/config.py]: Sorry about the force push!! I made a rebase with main to solve the conflicts.

File: vllm/engine/llm_engine.py [hunk 4]
```diff
@@ -254,43 +254,27 @@ def __init__(
...
-            model_config.override_neuron_config,
-            model_config.rope_scaling,
-            model_config.rope_theta,
-            model_config.tokenizer_revision,
-            model_config.trust_remote_code,
-            model_config.dtype,
-            model_config.max_model_len,
-            load_config.download_dir,
-            load_config.load_format,
+            "mm_processor_kwargs=%s, pooling_config=%s)", VLLM_VERSION,
```
Comment (maxdebayser) [inline vllm/engine/llm_engine.py:257]: It looks like the linter did more than it should have here. Ideally a commit should not change code lines that are unrelated.

File: vllm/model_executor/layers/pooler.py [hunk 5]
```diff
@@ -13,6 +14,33 @@ class PoolingType(IntEnum):
...
+
+    def __init__(self, pooling_type: str, normalize: bool):
+        self.pooling_type = self.get_pooling_type(pooling_type)
+        self.normalize = normalize
+
+    def get_pooling_type(self, pooling_type_name: str) -> PoolingType:
+        pooling_types = PoolingType.__dict__.items()
+        return PoolingType(
+            next((value for key, value in pooling_types
+                  if key.lower() in pooling_type_name), 2))
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:43]:
```suggestion
                  if key.lower() in pooling_type_name), PoolingType.CLS))
```

File: vllm/transformers_utils/tokenizer_group/__init__.py [hunk 6]
```diff
@@ -23,7 +23,8 @@ def init_tokenizer_from_configs(model_config: ModelConfig,
                        max_input_length=None,
                        tokenizer_mode=model_config.tokenizer_mode,
                        trust_remote_code=model_config.trust_remote_code,
-                       revision=model_config.tokenizer_revision)
+                       revision=model_config.tokenizer_revision,
+                       do_lower_case=model_config.do_lower_case)
```
Comment (maxdebayser) [inline vllm/transformers_utils/tokenizer_group/__init__.py:27]: Can we not add do_lower_case to the dict if it's not set? 

```suggestion
                       revision=model_config.tokenizer_revision)

    if model_config.do_lower_case is not None:
        init_kwargs["do_lower_case"] = model_config.do_lower_case
```

I think it's better if `"do_lower_case" in init_kwargs` evaluates to False if this parameter is not set

File: vllm/config.py [hunk 7]
```diff
@@ -1835,6 +1851,9 @@ def _get_and_verify_max_len(
                 raise ValueError(
                     f"{msg} To allow overriding this maximum, set "
                     "the env var VLLM_ALLOW_LONG_MAX_MODEL_LEN=1")
+
+    if bert_config:
```
Comment (maxdebayser) [inline vllm/config.py:1855]: Can we put this at line 1821 like this?

```python
if bert_config and "max_seq_lenght" in bert_config:
    derived_max_model_len = bert_config["max_seq_length"]
```

Otherwise it will disable the user override logic.

File: vllm/config.py [hunk 8]
```diff
@@ -173,6 +175,9 @@ def __init__(self,
                                     code_revision, rope_scaling, rope_theta,
                                     config_format)
         self.hf_text_config = get_hf_text_config(self.hf_config)
+        self.pooling_config = self.get_pooling_config()
+        self.bert_config = self._get_bert_config()
+        self.do_lower_case = self._get_bert_tokenization_config()
```
Comment (maxdebayser) [inline vllm/config.py:180]:
```suggestion
        self.bert_config, self.do_lower_case = self._get_bert_config()
```

File: vllm/config.py [hunk 9]
```diff
@@ -1797,6 +1820,9 @@ def _get_and_verify_max_len(
                     "original_max_position_embeddings"]
             derived_max_model_len *= scaling_factor
 
+    if bert_config and "max_seq_lenght" in bert_config:
```
Comment (maxdebayser) [inline vllm/config.py:1823]: My bad:
```suggestion
    if bert_config and "max_seq_length" in bert_config:
```

File: vllm/transformers_utils/config.py [hunk 10]
```diff
@@ -232,6 +235,102 @@ def get_config(
...
+    - model (str): The name of the Sentence Transformer 
+    BERT model.
+    - revision (str, optional): The revision of the m
+    odel to use. Defaults to 'main'.
+
+    Returns:
+    - dict: A dictionary containing the configuration parameters 
+    for the Sentence Transformer BERT model.
+    """
+    bert_dict = get_hf_file_to_dict("sentence_bert_config.json", model,
```
Comment (maxdebayser) [inline vllm/transformers_utils/config.py:323]: "sentence_bert_config.json" is one of several possible names: https://github.com/UKPLab/sentence-transformers/blob/f286d9f210824d6ea1563e789f49894b19c24f0e/sentence_transformers/models/Transformer.py#L421C13-L421C24

File: vllm/transformers_utils/config.py [hunk 11]
```diff
@@ -232,6 +235,102 @@ def get_config(
...
+        pooling_dict = get_hf_file_to_dict(pooling_file_name, model, revision)
+        pooling_type_name = next(
+            (item for item, val in pooling_dict.items() if val is True), None)
+
+        return {"pooling_type": pooling_type_name, "normalize": normalize}
+
+    return None
+
+
+def get_sentence_transformer_bert_config(model, revision='main'):
```
Comment (maxdebayser) [inline vllm/transformers_utils/config.py:308]: Maybe rename to `get_sentence_transformer_tokenizer_config`

File: vllm/transformers_utils/tokenizer_group/__init__.py [hunk 12]
```diff
@@ -25,6 +25,9 @@ def init_tokenizer_from_configs(model_config: ModelConfig,
                        trust_remote_code=model_config.trust_remote_code,
                        revision=model_config.tokenizer_revision)
 
+    if model_config.do_lower_case is not None:
+        init_kwargs["do_lower_case"] = model_config.do_lower_case
+
```
Comment (maxdebayser) [inline vllm/transformers_utils/tokenizer_group/__init__.py:30]:
```suggestion
    if model_config.bert_config is not None and "do_lower_case" in model_config.bert_config:
        init_kwargs["do_lower_case"] = model_config.bert_config["do_lower_case"]
```

File: vllm/transformers_utils/config.py (hunk 10 above)
Comment (flaviabeo) [reply vllm/transformers_utils/config.py:323]: Thank you! Will be adding these.

File: vllm/config.py [hunk 13]
```diff
@@ -1734,6 +1755,7 @@ def _get_and_verify_max_len(
         "max_seq_length",
         "seq_len",
     ]
+
```
Comment (maxdebayser) [inline vllm/config.py:1758]: nit: added whitespace

File: vllm/config.py [hunk 14]
```diff
@@ -1829,6 +1854,7 @@ def _get_and_verify_max_len(
                 raise ValueError(
                     f"{msg} To allow overriding this maximum, set "
                     "the env var VLLM_ALLOW_LONG_MAX_MODEL_LEN=1")
+
```
Comment (maxdebayser) [inline vllm/config.py:1857]: nit: added whitespace

File: vllm/model_executor/models/llama.py [hunk 15]
```diff
@@ -271,14 +271,13 @@ def forward(
...
 
-    def __init__(
-        self,
-        config: LlamaConfig,
-        cache_config: Optional[CacheConfig] = None,
-        quant_config: Optional[QuantizationConfig] = None,
-        lora_config: Optional[LoRAConfig] = None,
-        prefix: str = "",
-    ) -> None:
+    def __init__(self,
```
Comment (maxdebayser) [inline vllm/model_executor/models/llama.py:274]: I know this is the linter acting up, but can we try to commit without this formatting change?

File: vllm/model_executor/layers/pooler.py [hunk 16]
```diff
@@ -13,6 +14,33 @@ class PoolingType(IntEnum):
...
+
+    def __init__(self, pooling_type: str, normalize: bool):
+        self.pooling_type = self.get_pooling_type(pooling_type)
+        self.normalize = normalize
+
+    def get_pooling_type(self, pooling_type_name: str) -> PoolingType:
+        pooling_types = PoolingType.__dict__.items()
+        return PoolingType(
+            next((value for key, value in pooling_types
+                  if key.lower() in pooling_type_name), PoolingType.CLS))
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:43]: To better integrate with #9697, can we allow the user to override this via CLI? The order of priority should be: `EngineArgs` > sentence-transformers > model defaults

File: vllm/engine/arg_utils.py [hunk 17]
```diff
@@ -850,13 +852,29 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
             'priority (lower value means earlier handling) and time of '
             'arrival deciding any ties).')
 
+        parser.add_argument('--pooling-type',
+                            type=str,
+                            default='CLS',
+                            choices=['LAST', 'ALL', 'CLS', 'MEAN'],
```
Comment (maxdebayser) [inline vllm/engine/arg_utils.py:858]: Let's generate this list automatically from the enum

File: vllm/engine/arg_utils.py [hunk 18]
```diff
@@ -850,13 +852,29 @@ def add_cli_args(parser: FlexibleArgumentParser) -> FlexibleArgumentParser:
...
+                            help='Wheter to normalize the pooled data.')
+
         return parser
 
     @classmethod
     def from_cli_args(cls, args: argparse.Namespace):
         # Get the list of attributes of this dataclass.
         attrs = [attr.name for attr in dataclasses.fields(cls)]
         # Set the attributes from the parsed arguments.
+        pooling_config = PoolingConfig(pooling_type=args.pooling_type,
```
Comment (maxdebayser) [inline vllm/engine/arg_utils.py:874]: Isn't `pooling_type` an `str` here? Perhaps a conversion to enum type is missing.

File: vllm/model_executor/layers/pooler.py [hunk 19]
```diff
@@ -32,7 +34,7 @@ class PoolingConfig:
         type enum value corresponding to the given string.
     """
 
-    def __init__(self, pooling_type: str, normalize: bool):
+    def __init__(self, pooling_type: str, normalize: Optional[bool] = None):
```
Comment (maxdebayser) [inline vllm/model_executor/layers/pooler.py:37]: Is this just to satisfy `mypy`? I think this object should not be constructed with `normalize == None`.

File: vllm/config.py [hunk 20]
```diff
@@ -179,7 +180,7 @@ def __init__(self,
                                     code_revision, rope_scaling, rope_theta,
                                     config_format)
         self.hf_text_config = get_hf_text_config(self.hf_config)
-        self.pooling_config = self.get_pooling_config()
+        self.pooling_config = pooling_config or self.get_pooling_config()
```
Comment (maxdebayser) [inline vllm/config.py:183]: Since the two new flags can be used independently, I think we need to be more careful here. The flags are used to override whatever is loaded from the config files. By putting both options in the pooling_config, we risk overriding `normalize` if only `pooling-type` is set and vice-versa.

File: vllm/model_executor/layers/pooler.py (hunk 16 above)
Comment (flaviabeo) [reply vllm/model_executor/layers/pooler.py]: Working on it!

File: vllm/config.py [hunk 21]
```diff
@@ -419,12 +420,23 @@ def _verify_bnb_config(self) -> None:
                 "fallback to the eager mode.")
             self.enforce_eager = True
 
-    def get_pooling_config(self) -> Optional[PoolingConfig]:
+    def get_pooling_config(self, pooling_type_arg,
```
Comment (maxdebayser) [inline vllm/config.py:423]: Can we add type annotations here?

**Comment by DarkLight1337** [comment]
Sorry, can you hold this for a bit? I'm trying to get #9697 merged which will definitely introduce merge conflicts. The CI is broken right now so we are all waiting for force merge.

File: vllm/config.py [hunk 22]
```diff
@@ -107,6 +108,8 @@ class ModelConfig:
             can not be gathered from the vllm arguments.
         config_format: The config format which shall be loaded.
             Defaults to 'auto' which defaults to 'hf'.
+        bert_config: tokenizationconfiguration dictionary for a given
```
Comment (maxdebayser) [inline vllm/config.py:111]: `bert_config` is not actually a parameter of the `__init__` method. The documentation can be moved close to the initialization of `self.bert_config` or removed. Otherwise the docstring will be wrong.

**Comment by maxdebayser** [comment]
In [vllm/engine/arg_utils.py](https://github.com/vllm-project/vllm/blob/b63c64d95b01cc955a56bba37d055ad36aa81abd/vllm/engine/arg_utils.py#L853) the `MEAN` type is missing in the set of acceptable values for the command line arg:
```
        parser.add_argument(
            '--pooling-type',
            choices=['LAST', 'ALL', 'CLS', 'STEP'],
            default=None,
```
In before the rebase you had a version where this list was generated automatically from the Enum type. Can you add this again?

File: vllm/config.py [hunk 23]
```diff
@@ -271,12 +278,28 @@ def _init_pooler_config(
...
+                pooling_config_from_file = PoolerConfig(
+                    pooling_type=pooling_type_from_file,
+                    pooling_norm=normalize_from_file,
+                    pooling_softmax=pooling_softmax,
+                    pooling_step_tag_id=pooling_step_tag_id,
+                    pooling_returned_token_ids=pooling_returned_token_ids)
+                if pooling_type is not None:
+                    pooling_config_from_file.pooling_type = pooling_type
+                if pooling_norm is not None:
+                    pooling_config_from_file.pooling_norm = pooling_norm
```
Comment (DarkLight1337) [inline vllm/config.py:294]: Can we clean this logic up a bit? In particular, we can directly assign the contents of `pooling_config` to the input variables before constructing `PoolerConfig`.

Comment (flaviabeo) [reply vllm/config.py]: Sure, let me know if I can improve more!

**Review by maxdebayser — APPROVED** [review]
LGTM

File: vllm/transformers_utils/config.py [hunk 24]
```diff
@@ -232,6 +235,157 @@ def get_config(
     return config
 
 
+def get_hf_file_to_dict(file_name,
+                        model,
+                        revision,
+                        token: Optional[str] = None):
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py:241]: Can we add type annotations here?

File: vllm/transformers_utils/config.py [hunk 25]
```diff
@@ -232,6 +235,157 @@ def get_config(
...
+                    EntryNotFoundError, LocalEntryNotFoundError) as e:
+                logger.debug("File or repository not found in hf_hub_download",
+                             e)
+                return None
+            file_path = Path(hf_hub_file)
+
+        with open(file_path, "r") as file:
+            config_dict = json.load(file)
+
+        return config_dict
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py:278]:
```suggestion
            return json.load(file)
```

File: vllm/model_executor/layers/pooler.py [hunk 26]
```diff
@@ -27,7 +28,7 @@ class Pooler(nn.Module):
     3. Returns structured results as `PoolerOutput`.
 
     Attributes:
-        pooling_type: The type of pooling to use (LAST, ALL, CLS).
+        pooling_type: The type of pooling to use (LAST, ALL, CLS, MEAN).
```
Comment (DarkLight1337) [inline vllm/model_executor/layers/pooler.py:31]:
```suggestion
        pooling_type: The type of pooling to use.
```
To avoid having to keep updating, let's just remove this. The type annotation should be informative enough.

File: vllm/config.py [hunk 27]
```diff
@@ -271,9 +278,20 @@ def _init_pooler_config(
...
+            normalize_ = pooling_norm
+            pooling_config = get_pooling_config(self.model, self.revision)
+            if pooling_config is not None:
+                pooling_type_ = pooling_config["pooling_type"]
+                normalize_ = pooling_config["normalize"]
+                # override if user specifies pooling_type and/or pooling_norm
+                if pooling_type is not None:
+                    pooling_type_ = pooling_type
+                if pooling_norm is not None:
+                    normalize_ = pooling_norm
```
Comment (DarkLight1337) [inline vllm/config.py:291]:
```suggestion
            pooling_config = get_pooling_config(self.model, self.revision)
            if pooling_config is not None:
                # override if user specifies pooling_type and/or pooling_norm
                if pooling_config["pooling_type"] is not None:
                    pooling_type = pooling_config["pooling_type"]
                if pooling_config["normalize"]  is not None:
                    normalize = pooling_config["normalize"]
```

File: vllm/transformers_utils/config.py [hunk 28]
```diff
@@ -232,6 +235,157 @@ def get_config(
...
+
+        if pooling_type_name is not None:
+            pooling_type_name = get_pooling_config_name(pooling_type_name)
+
+        return {"pooling_type": pooling_type_name, "normalize": normalize}
+
+    return None
+
+
+def get_pooling_config_name(pooling_name):
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py:329]:
```suggestion
def get_pooling_config_name(pooling_name: str) -> str:
```

File: vllm/transformers_utils/config.py [hunk 29]
```diff
@@ -232,6 +235,157 @@ def get_config(
...
+        if pooling_type_name in supported_pooling_types:
+            return pooling_type_name
+    except NotImplementedError as e:
+        logger.debug("Pooling type not supported", e)
+        return None
+
+
+def get_sentence_transformer_tokenizer_config(model,
+                                              revision='main',
+                                              token: Optional[str] = None):
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py:352]:
```suggestion
def get_sentence_transformer_tokenizer_config(model: str,
                                              revision: str = 'main',
                                              token: Optional[str] = None):
```

File: vllm/transformers_utils/config.py [hunk 30]
```diff
@@ -294,20 +448,15 @@ def _reduce_modelconfig(mc: ModelConfig):
             exc_info=e)
 
 
-def load_params_config(model, revision) -> PretrainedConfig:
+def load_params_config(model,
+                       revision,
+                       token: Optional[str] = None) -> PretrainedConfig:
```
Comment (DarkLight1337) [inline vllm/transformers_utils/config.py:453]:
```suggestion
def load_params_config(model: str,
                       revision: str,
                       token: Optional[str] = None) -> PretrainedConfig:
```

File: vllm/config.py (hunk 27 above)
Comment (flaviabeo) [reply vllm/config.py:291]: This way, the pooling config from model's files will override the user input. Is that the expected behavior for this?

Comment (DarkLight1337) [reply vllm/config.py:291]: Oops, let me edit this.

Comment (DarkLight1337) [reply vllm/config.py:291]: Edited.

Comment (flaviabeo) [reply vllm/config.py:291]: Got it! Thank you!

File: vllm/model_executor/layers/pooler.py (hunk 26 above)
Comment (DarkLight1337) [reply vllm/model_executor/layers/pooler.py:31]: This still hasn't been updated yet.

Comment (flaviabeo) [reply vllm/model_executor/layers/pooler.py]: Sorry! Done.

**Review by DarkLight1337 — APPROVED** [review]
Thanks for your patience!

File: vllm/config.py [hunk 31]
```diff
@@ -186,6 +187,7 @@ def __init__(
                                     code_revision, rope_scaling, rope_theta,
                                     config_format)
         self.hf_text_config = get_hf_text_config(self.hf_config)
+        self.bert_config = self._get_bert_config()
```
Comment (mgoin) [inline vllm/config.py:190]: Could we call this something more general than "bert_config"? Maybe "encoder_config"?

**Comment by flaviabeo** [comment]
I was able to reproduce the error we are seeing at the CI tests in my environment - it is regarding another change that was included recently. The two environments compared in this test are:
```sh
env 1  {'VLLM_TORCH_COMPILE_LEVEL': '0'}
env 2  {'VLLM_TORCH_COMPILE_LEVEL': '3'}
```
All the vllm engine args and model configuration are the same, but when the test compares the tensors produced with this two different torch compiles, it fails. The vectors are different.

Following are the engine args:
```sh
['--enforce-eager', '--task', 'embedding', '-pp', '1', '-tp', '1', '--load-format', 'dummy']
```
And model config parameters:
```sh
model='BAAI/bge-multilingual-gemma2', speculative_config=None, tokenizer='BAAI/bge-multilingual-gemma2', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=8192, download_dir=None, load_format=LoadFormat.DUMMY, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=None, enforce_eager=True, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=BAAI/bge-multilingual-gemma2, num_scheduler_steps=1, chunked_prefill_enabled=False multi_step_stream_outputs=True, enable_prefix_caching=False, use_async_output_proc=False, use_cached_outputs=True, chat_template_text_format=string, mm_processor_kwargs=None, pooler_config=PoolerConfig(pooling_type='LAST', pooling_norm=False, pooling_softmax=None, pooling_step_tag_id=None, pooling_returned_token_ids=None))
```

**Comment by DarkLight1337** [comment]
@youkaichao maybe there is a bug in using `torch.compile`?

**Comment by youkaichao** [comment]
the mse is `mse=0.0005555402021855116` , is it acceptable? `torch.compile` can indeed change the model's output slightly.

**Comment by DarkLight1337** [comment]
It is acceptable. I think we should measure cosine similarity instead of MSE for embedding models though.

**Comment by youkaichao** [comment]
> It is acceptable. I think we should measure cosine similarity instead of MSE for embedding models though.

feel free to update the tests then.
